### ISSUE TITLE
A soft memory limit for `clickhouse-keeper`

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#56348]
**Use case**

If the memory usage of `clickhouse-keeper` is above a threshold, it should refuse operations that increase the memory consumption: `create` operations, `multi`-transactions involving `create`, and possibly `set`.

**Describe the solution you'd like**

We can set it as 80% by default.

**Comment by hanfei1991** [comment]
@antonio2368  should we only limit leader? because follower may consume more memory for example apply snapshot

**Comment by antonio2368** [comment]
@hanfei1991 no, we should limit on each replica before the request even get to RAFT.
When we are in Raft it's much harder to properly handle OOM cases. We already do have some form of limits in place where we throw OutOfMemory exceptions and if we get that exception anywhere inside Raft we currently abort which is okay IMO (it should be the same thing in ZK when it goes over heap size).
